File: app/src/main/java/name/monwf/customiuizer/utils/ResolveInfoAdapter.java
package name.monwf.customiuizer.utils;

import java.util.ArrayList;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;

import android.content.Context;
import android.content.pm.PackageManager;
import android.content.pm.ResolveInfo;
import android.graphics.Bitmap;
import android.graphics.drawable.Drawable;
import android.graphics.drawable.TransitionDrawable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.Filter;
import android.widget.Filterable;
import android.widget.ImageView;
import android.widget.TextView;

import name.monwf.customiuizer.R;

public class ResolveInfoAdapter extends BaseAdapter implements Filterable {

	private final Context ctx;
	private final PackageManager pm;
	private final LayoutInflater mInflater;
	private final ThreadPoolExecutor pool;
	private final ItemFilter mFilter = new ItemFilter();
	private final CopyOnWriteArrayList<ResolveInfo> originalAppList = new CopyOnWriteArrayList<ResolveInfo>();
	private final CopyOnWriteArrayList<ResolveInfo> filteredAppList = new CopyOnWriteArrayList<ResolveInfo>();

	public ResolveInfoAdapter(Context context, ArrayList<ResolveInfo> arr) {
		ctx = context;
		pm = ctx.getPackageManager();
		mInflater = LayoutInflater.from(context);
		originalAppList.addAll(arr);
		filteredAppList.addAll(arr);
		int cpuCount = Runtime.getRuntime().availableProcessors();
		pool = new ThreadPoolExecutor(cpuCount + 1, cpuCount * 2 + 1, 2, TimeUnit.SECONDS, new LinkedBlockingQueue<Runnable>());
	}

	public int getCount() {
		return filteredAppList.size();
	}

	public ResolveInfo getItem(int position) {
		return filteredAppList.get(position);
	}

	public long getItemId(int position) {
		return position;
	}

	public View getView(final int position, View convertView, ViewGroup parent) {
		View row;
		if (convertView != null)
			row = convertView;
		else
			row = mInflater.inflate(R.layout.applist_item11, parent, false);

		ImageView itemIsDis = row.findViewById(R.id.icon_disable);
		TextView itemTitle = row.findViewById(android.R.id.title);
		ImageView itemIcon = row.findViewById(android.R.id.icon);

		ResolveInfo ri = getItem(position);
		itemIcon.setTag(position);

		AppData ad = new AppData();
		ad.pkgName = ri.activityInfo.applicationInfo.packageName;
		ad.actName = ri.activityInfo.name;
		ad.enabled = ri.activityInfo.enabled;
		ad.label = ri.loadLabel(pm).toString();

		itemTitle.setText(ad.label);
		itemIsDis.setVisibility(ad.enabled ? View.INVISIBLE : View.VISIBLE);
		Bitmap icon = Helpers.memoryCache.get(ad.pkgName + "|" + ad.actName);

		if (icon == null) {
			Drawable[] dualIcon = new Drawable[1];
			dualIcon[0] = ctx.getResources().getDrawable(R.drawable.card_icon_default, ctx.getTheme());
			TransitionDrawable crossfader = new TransitionDrawable(dualIcon);
			crossfader.setCrossFadeEnabled(true);
			itemIcon.setImageDrawable(crossfader);
			(new BitmapCachedLoader(itemIcon, ad, ctx)).executeOnExecutor(pool);
		} else {
			itemIcon.setImageBitmap(icon);
		}

		//row.setEnabled(true);
		return row;
	}

	private class ItemFilter extends Filter {
		@Override
		protected FilterResults performFiltering(CharSequence constraint) {
			String filterString = constraint.toString().toLowerCase();
			FilterResults results = new FilterResults();

			int count = originalAppList.size();
			final ArrayList<ResolveInfo> nlist = new ArrayList<ResolveInfo>();
			ResolveInfo filterableData;

			for (int i = 0; i < count; i++) {
				filterableData = originalAppList.get(i);
				if (filterableData.loadLabel(pm).toString().toLowerCase().contains(filterString)) nlist.add(filterableData);
			}

			results.values = nlist;
			results.count = nlist.size();
			return results;
		}

		@Override
		@SuppressWarnings("unchecked")
		protected void publishResults(CharSequence constraint, FilterResults results) {
			filteredAppList.clear();
			if (results.count > 0 && results.values != null)
			filteredAppList.addAll((ArrayList<ResolveInfo>)results.values);
			notifyDataSetChanged();
		}
	}

	@Override
	public Filter getFilter() {
		return mFilter;
	}
}

File: app/src/main/java/tv/withaibuild/customiuizer/utils/ResolveInfoAdapter.kt
package tv.withaibuild.customiuizer.utils

import android.content.Context
import android.content.pm.PackageManager
import android.content.pm.ResolveInfo
import android.graphics.drawable.TransitionDrawable
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.Filter
import android.widget.Filterable
import android.widget.ImageView
import android.widget.TextView
import tv.withaibuild.customiuizer.R
import java.util.ArrayList
import java.util.Locale
import java.util.concurrent.CopyOnWriteArrayList

class ResolveInfoAdapter(context: Context, arr: ArrayList<ResolveInfo>) : BaseAdapter(), Filterable {

    private val ctx: Context = context
    private val pm: PackageManager = ctx.packageManager
    private val inflater: LayoutInflater = LayoutInflater.from(context)
    private val filter = ItemFilter()
    private val originalAppList = CopyOnWriteArrayList(arr)
    private val filteredAppList = CopyOnWriteArrayList(arr)

    override fun getCount(): Int = filteredAppList.size

    override fun getItem(position: Int): ResolveInfo = filteredAppList[position]

    override fun getItemId(position: Int): Long = position.toLong()

    override fun getView(position: Int, convertView: View?, parent: ViewGroup?): View {
        val holder = (convertView?.tag as? ViewHolder) ?: run {
            val row = inflater.inflate(R.layout.applist_item11, parent, false)
            ViewHolder(row).also { row.tag = it }
        }

        val ri = getItem(position)
        holder.icon.tag = position

        val ad = AppData().apply {
            pkgName = ri.activityInfo.applicationInfo.packageName
            actName = ri.activityInfo.name
            enabled = ri.activityInfo.enabled
            label = ri.loadLabel(pm).toString()
        }

        holder.title.text = ad.label
        holder.disableIcon.visibility = if (ad.enabled) View.INVISIBLE else View.VISIBLE
        val icon = Helpers.memoryCache[ad.pkgName + "|" + ad.actName]

        if (icon == null) {
            val dualIcon = arrayOf(ctx.resources.getDrawable(R.drawable.card_icon_default, ctx.theme))
            val crossfader = TransitionDrawable(dualIcon)
            crossfader.setCrossFadeEnabled(true)
            holder.icon.setImageDrawable(crossfader)
            BitmapCachedLoader(holder.icon, ad, ctx).execute()
        } else {
            holder.icon.setImageBitmap(icon)
        }

        return holder.root
    }

    private class ViewHolder(val root: View) {
        val disableIcon: ImageView = root.findViewById(R.id.icon_disable)
        val title: TextView = root.findViewById(android.R.id.title)
        val icon: ImageView = root.findViewById(android.R.id.icon)
    }

    private inner class ItemFilter : Filter() {
        override fun performFiltering(constraint: CharSequence?): FilterResults {
            val filterString = constraint?.toString()?.lowercase(Locale.ROOT) ?: ""
            val results = FilterResults()
            val nlist = ArrayList<ResolveInfo>()

            for (ri in originalAppList) {
                if (ri.loadLabel(pm).toString().lowercase(Locale.ROOT).contains(filterString)) {
                    nlist.add(ri)
                }
            }

            results.values = nlist
            results.count = nlist.size
            return results
        }

        @Suppress("UNCHECKED_CAST")
        override fun publishResults(constraint: CharSequence?, results: FilterResults?) {
            filteredAppList.clear()
            if (results != null && results.count > 0 && results.values != null) {
                filteredAppList.addAll(results.values as ArrayList<ResolveInfo>)
            }
            notifyDataSetChanged()
        }
    }

    override fun getFilter(): Filter = filter
}
